### PR TITLE
Substructure to Shared Library

### PR DESCRIPTION
**Pull request by jackaraz**
**Context:**
This PR includes code flow optimization and compilation efficiency updates. 

* The substructure module has been externalized as a shared library. This allows faster compilation time for individual projects.
* HEPTopTagger has been added to the workflow. The main code will be downloaded from the latest release of [this repo](https://github.com/MadAnalysis/HEPTopTagger).

PR also includes additional functionality to detect fastjet contrib and HTT to compile MadAnalysis accordingly.

### CODE REVIEW

File: madanalysis/install/install_fastjetcontrib.py [hunk 1]
```diff
@@ -85,7 +88,9 @@ def Unpack(self):
 
     def Configure(self):
         # Input
-        theCommands=['./configure','--fastjet-config='+self.bindir]
+        # TODO: figure out how to give `-std=c++11 -fPIC` together to CXXFLAGS
+        # using " or ' doesn't work on linux systems
+        theCommands = ['./configure', '--fastjet-config=' + self.bindir, 'CXXFLAGS=-fPIC']
```
Comment (jackaraz) [inline madanalysis/install/install_fastjetcontrib.py:93]: **TODO:** fjcontrib still needs to be compiled with `CXXFLAGS="-std=c++11 -fPIC"`; however couldn't figure out how to feed them together to the terminal. The configuration file doesn't accept `"` or `'` through the python interface, although both work when done by hand.

**Review by jackaraz — COMMENTED** [review]
Regarding linking errors in Linux systems.

**Comment by BFuks** [comment]
Hi Jack. Can you please first resolve the conflicts before I start reviewing this branch? Thanks in advance!

**Comment by jackaraz** [comment]
> Hi Jack. Can you please first resolve the conflicts before I start reviewing this branch? Thanks in advance!

The conflict has been resolved.

**Comment by BFuks** [comment]
> > Hi Jack. Can you please first resolve the conflicts before I start reviewing this branch? Thanks in advance!
> 
> The conflict has been resolved.

Great. Starting the review!

File: madanalysis/install/install_fastjetcontrib.py (hunk 1 above)
Comment (BFuks) [reply madanalysis/install/install_fastjetcontrib.py:93]: Have you tried with ```\'```?

Comment (jackaraz) [reply madanalysis/install/install_fastjetcontrib.py:86]: yeap didn't work for me even tried the Unicode string :/

File: tools/SampleAnalyzer/Interfaces/HEPTopTagger/HTT.h [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+                //top mass range
+                MAfloat32 top_mass = 172.3;
+                MAfloat32 W_mass = 80.4;
+                MAfloat32 Mtop_min = 150.;
+                MAfloat32 Mtop_max = 200.; //set_top_range(min,max)
+
+                // set top mass ratio range
+                MAfloat32 fw = 0.15;
+                MAfloat32 mass_ratio_range_min = (1-fw)*80.379/172.9;
```
Comment (BFuks) [inline tools/SampleAnalyzer/Interfaces/HEPTopTagger/HTT.h:85]: We need a function that update this when the user change the mass of the top or the W boson.

File: tools/SampleAnalyzer/Interfaces/substructure/ClusterBase.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+#include "SampleAnalyzer/Commons/Service/LogService.h"
+#include "SampleAnalyzer/Interfaces/substructure/Commons.h"
+
+namespace fastjet
+{
+    class JetDefinition;
+    class PseudoJet;
+}
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (jackaraz) [reply tools/SampleAnalyzer/Interfaces/substructure/ClusterBase.h:46]: nope, not needed indeed.

File: tools/SampleAnalyzer/Interfaces/substructure/README.md [hunk 4]
```diff
@@ -1,39 +1,51 @@
...
+is designed to work with a single jet, whereas the latter is designed to work
+with a collection of jets. The precise action of the `Execute` function depends
+on the substructure tools to be used. The `Execute` function, as suggested by
+its name, needs to be executed during the analysis execution.
 
 ## Outline
 * [Examples](#examples)
   * [SoftDrop](#softdrop)
-  * [Filter](#filter)
+  * [Filtering](#filter)
```
Comment (jackaraz) [inline tools/SampleAnalyzer/Interfaces/substructure/README.md:34]: `(#filter) ` -> `(#filtering)`

File: tools/SampleAnalyzer/Interfaces/substructure/README.md [hunk 5]
```diff
@@ -1,39 +1,51 @@
...
   * [Energy Correlator](#energy-correlator)
   * [Nsubjettiness](#nsubjettiness)
-  * [Cluster](#cluster)
+  * [Clustering](#cluster)
```
Comment (jackaraz) [inline tools/SampleAnalyzer/Interfaces/substructure/README.md:37]: `(#cluster)` -> `(#clustering)`

File: tools/SampleAnalyzer/Interfaces/substructure/README.md [hunk 6]
```diff
@@ -1,39 +1,51 @@
...
   * [VariableR](#variabler)
-  * [Pruner](#pruner)
-  * [Recluster](#recluster)
+  * [Pruning](#pruner)
```
Comment (jackaraz) [inline tools/SampleAnalyzer/Interfaces/substructure/README.md:39]: `(#pruner)` -> `(#pruning)`

File: tools/SampleAnalyzer/Interfaces/substructure/README.md [hunk 7]
```diff
@@ -1,39 +1,51 @@
...
+  * [Reclustering](#recluster)
```
Comment (jackaraz) [inline tools/SampleAnalyzer/Interfaces/substructure/README.md:40]: `(#recluster)` -> `(#reclustering)`

File: tools/SampleAnalyzer/Interfaces/substructure/README.md [hunk 8]
```diff
@@ -69,16 +84,19 @@ of jet definitions. All available `jet_algorithm` definitions are shown with the
 |      `cdfjetclu`      | `radius=0.4`, `ptmin=5.`, `overlap=0.5`, `seed=1.`, `iratch=0.`                    |
 |     `cdfmidpoint`     | `radius=0.4`, `ptmin=5.`, `overlap=0.5`, `seed=1.`, `iratch=0.`, `areafraction=1.` |
 |       `siscone`       | `radius=0.4`, `ptmin=5.`, `overlap=0.5`, `input_ptmin=5.`, `npassmax=1.`           |
-|      `VariableR`      | `rho=2000.`, `minR=0.`, `maxR=2.`, `ptmin=20.` `exclusive=False` `clustertype=CALIKE` `strategy=Best` |
+|      `VariableR`      | `rho=2000.`, `minR=0.`, `maxR=2.`, `ptmin=20.` `exclusive=False`                   |
+|                       | `clustertype=CALIKE` `strategy=Best`                                               |
```
Comment (jackaraz) [inline tools/SampleAnalyzer/Interfaces/substructure/README.md:88]: separating a table line like this creates a separate line, please let markdown take care of the line handling
```
|      `VariableR`      | `rho=2000.`, `minR=0.`, `maxR=2.`, `ptmin=20.` `exclusive=False`                   |
 |                       | `clustertype=CALIKE` `strategy=Best`                                               |
``` 
needs to be 
```
|      `VariableR`      | `rho=2000.`, `minR=0.`, `maxR=2.`, `ptmin=20.` `exclusive=False` `clustertype=CALIKE` `strategy=Best` |
```

**Review by jackaraz — COMMENTED** [review]
regarding `README.md`

**Review by BFuks — APPROVED** [review]

**Pull request merged**